### PR TITLE
Creating forgot password feature

### PR DESCRIPTION
**Pull request by Abed01-lab**
added nodemailer config,
added env variables

### CODE REVIEW

**Review by OwaisAD — COMMENTED** [review]
There is a flaw in that someone can call (from postman or playground)  verifyUser mutation and then resetPassword mutation ACTUALLY skipping the verifyCode mutation, which means we have to find a fix for validating.

Other than that couple things here and there - I like what I see generally 😍 Good stuff

**Comment by Abed01-lab** [comment]
> There is a flaw in that someone can call (from postman or playground) verifyUser mutation and then resetPassword mutation ACTUALLY skipping the verifyCode mutation, which means we have to find a fix for validating.
> 
> Other than that couple things here and there - I like what I see generally 😍 Good stuff

Good spot! We should definitely find a way to fix it

File: client/src/pages/forgot-password/EmailStep.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import { ZodError } from "zod";
+import toast from "react-hot-toast";
+import { emailSchema } from "../../validators/emailSchema";
+import { useMutation } from "@apollo/client";
+import { verifyUser } from "../../graphql/mutations/user.mutations";
+
+interface EmailStepProps extends StepProps {
+  setEmail: Dispatch<SetStateAction<string>>;
+  email: string;
+}
```
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/EmailStep.tsx:19]: Yeah I dont agree here, I think it makes more sense to add a type if the type is used multiple places, but in this case the emailStepProps type is only used inside of the component, therefore not necessary to make a type and export it

File: client/src/pages/forgot-password/ForgotPassword.tsx
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/ForgotPassword.tsx:1]: We can invalidate the verificationCode when they leave the resetpassword or code verification step

[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/ForgotPassword.tsx:1]: we should talk about this

File: client/src/pages/forgot-password/EmailStep.tsx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+              <img
+                src="/logo.png"
+                alt="InvoPilot logo"
+                width={40}
+                height={40}
+                className="hover:scale-110 duration-150"
+              />
+              <Heading text="Forgot password?" />
+              <InputField
+                name="email"
```
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/EmailStep.tsx:68]: What ?

File: client/src/pages/forgot-password/VerificationStep.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import { verifyCode } from "../../graphql/mutations/user.mutations";
+import EaseInEaseOut from "../../components/framer_transitions/EaseInEaseOut";
+import { Card } from "../../components/Card";
+import { Heading } from "../../components/Heading";
+import { CircleProgress } from "../../components/CircleProgress";
+import { VerificationField } from "../../components/VerificationField";
+
+interface VerificationStepProps extends StepProps {
+  email: string;
+}
```
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/VerificationStep.tsx:17]: Same comment as abov

File: client/src/pages/forgot-password/VerificationStep.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      <Helmet>
+        <title>Forgot password - InvoPilot</title>
+      </Helmet>
+      <div className="min-h-screen w-full bg-gray-100 flex justify-center items-center">
+        <EaseInEaseOut>
+          <Card flex="col" justify="center" items="center">
+            <img src="/logo.png" alt="InvoPilot logo" width={40} height={40} className="hover:scale-110 duration-150" />
+            <Heading text="Enter 6-digit code?" />
+            <p className="text-center text-base leading-6 font-light text-gray-700">
+              The recovery code was sent to your email. Please enter the code:
```
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/VerificationStep.tsx:59]: Not a big fan of diverting off the figma design. When we are developing we shouldn't add extra to the design, only add what we have in the figma design

File: server/src/graphql/resolvers/user.resolver.ts [hunk 5]
```diff
@@ -1,9 +1,14 @@
 import { UserCreateAccountInput } from "../../types/user.types";
-import { createUserAndCompany } from "../../services/user.service";
+import { createUserAndCompany, resetPassword, verifyCode, verifyUser } from "../../services/user.service";
 
 export const userResolver = {
   Mutation: {
     createUserAndCompany: async (_parent: never, args: UserCreateAccountInput) => await createUserAndCompany(args),
+    verifyUser: async (_parent: never, args: { email: string }) => await verifyUser(args.email),
+    verifyCode: async (_parten: never, args: { email: string; verificationCode: string }) =>
+      await verifyCode(args.email, args.verificationCode),
```
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply server/src/graphql/resolvers/user.resolver.ts:9]: Should we make a resolver called auth and add all the login, Signup, ResetPassword, verifyUser and verifyCode ?

File: server/src/services/user.service.ts [hunk 6]
```diff
@@ -62,3 +69,84 @@ const createUser = async (userInput: UserInput) => {
...
+    if (error instanceof ZodError) {
+      const errorMessage = error.errors.map((err) => err.message).join(", ");
+      throw new Error(errorMessage);
+    }
+
+    throw error;
+  }
+};
+
+export const resetPassword = async (password: string, confirmPassword: string, email: string) => {
```
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply server/src/services/user.service.ts:126]: Here is a proposal for a solution:

We make the verifyCode mutation return a uuid, which we save in memory, we then need this uuid in the resetPassword step to check if the user is who he say he is.

File: server/src/tests/user.test.ts [hunk 7]
```diff
@@ -195,3 +198,262 @@ describe("Testing creating a user", () => {
...
+          verificationCode: code,
+        },
+      });
+
+    expect(response.body.data.verifyCode).toBe("Verification code is valid");
+  });
+});
+
+describe("Testing resetPassword method", () => {
+  it("Should should return unauthorized, if user tries to reset password without verification code", async () => {
```
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply server/src/tests/user.test.ts:326]: I added a proposal in the above comment

File: client/src/pages/forgot-password/EmailStep.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (OwaisAD) [reply client/src/pages/forgot-password/EmailStep.tsx:68]: In between the heading and password inputfield you can add a paragraph saying: "No worries, we will send you reset instructions."

File: client/src/pages/forgot-password/VerificationStep.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (OwaisAD) [reply client/src/pages/forgot-password/VerificationStep.tsx:59]: I have to disagree. I am a big fan of adding relevant stuff. Diverting from an old design scheme with good arguments is good. Therefore the above is just as relevant as adding the text beneath a password inputfield describing to the user what to do. Those are stuff we are missing in our mockup.

File: server/src/graphql/resolvers/user.resolver.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (OwaisAD) [reply server/src/graphql/resolvers/user.resolver.ts:9]: Sure, that could work.

File: server/src/services/user.service.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (OwaisAD) [reply server/src/services/user.service.ts:126]: Yes, that can work

File: client/src/pages/forgot-password/VerificationStep.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/VerificationStep.tsx:59]: Then the step should be change the design in figma, make it up to date, then request change in the source code, not vice versa

File: client/src/pages/forgot-password/EmailStep.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/EmailStep.tsx:68]: Again these things should be in figma first. Then we can change the source code

File: server/src/graphql/resolvers/user.resolver.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply server/src/graphql/resolvers/user.resolver.ts]: done

File: client/src/pages/forgot-password/ForgotPassword.tsx
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/ForgotPassword.tsx:1]: I have a fix

File: client/src/pages/forgot-password/ResetPassword.tsx [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+              <img
+                src="/logo.png"
+                alt="InvoPilot logo"
+                width={40}
+                height={40}
+                className="hover:scale-110 duration-150"
+              />
+              <Heading text="Create new password?" />
+              <p className="text-center text-base leading-6 font-light text-gray-700">
+                Your new password must be different from previously used passwords
```
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/ResetPassword.tsx:76]: done

File: client/src/pages/forgot-password/EmailStep.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Abed01-lab) [reply client/src/pages/forgot-password/EmailStep.tsx:68]: done, I had forgotten to add in the text

**Review by OwaisAD — APPROVED** [review]

**Pull request merged**